neural_compressor/adaptor/torch_utils/block_detector.py

```diff
@@ -0,0 +1,164 @@
+#
+# -*- coding: utf-8 -*-
+#
+# Copyright (c) 2021 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Block detector for Transformer-based model."""
+
+from ...utils.utility import LazyImport
+torch = LazyImport("torch")
+from typing import Dict, List, Union
+
+from ...utils import logger
+from .util import get_depth, get_dict_at_depth, get_element_under_depth
+
+BLOCK_PATTERNS = [
+    # [['OP_TYPE1', NUM_OPS], ['OP_TYPE2', NUM_OPS], ...]
+    [['Linear', 4], ['Linear', 4]], # TODO add model name
+    [['Linear', 2], ['Linear', 2]], # TODO add model name
+    [['Conv1D', 2], ['Conv1D', 2]], # GPT-2
+    [['Linear', 4], ['Linear', 3]], # Llama
+    [['Linear', 4], ['Linear', 2]], # T5-Encoder, OPT
+    [['Linear', 4], ['Linear', 1], ['Linear', 1]], # Bert
+    [['Linear', 4], ['Linear', 4], ['Linear', 2]],  # T5-Decoder
+]
+
+
+class TransformerModelBlockDetector:
+    """Detect the attention block and FFN block in transformer-based model."""
+    
+    def __init__(self, model: torch.nn.Module, pattern_lst: List[List[Union[str, int]]] = BLOCK_PATTERNS) -> None:
```

Comment:
it's weird. I think it will trigger issue. pls double confirm. torch.nn.Module is used to support Type Check. but where is the import torch? should use Lazy import to avoid framework dependency issue